// node-etcd/lib/index.js
// Generated by CoffeeScript 1.9.1
var Client, Etcd, HttpsAgent, Watcher, _, exports,
  bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

_ = require('underscore');

Watcher = require('./watcher');

Client = require('./client');

HttpsAgent = (require('https')).Agent;

Etcd = (function() {
  function Etcd(host, port, sslopts, client) {
    if (host == null) {
      host = '127.0.0.1';
    }
    if (port == null) {
      port = null;
    }
    if (sslopts == null) {
      sslopts = null;
    }
    if (client == null) {
      client = null;
    }
    this.watcher = bind(this.watcher, this);
    if (_.isArray(host)) {
      this.hosts = host;
      this.sslopts = port;
      this.client = sslopts;
    } else {
      if (port == null) {
        port = 4001;
      }
      this.hosts = [host + ":" + port];
      this.sslopts = sslopts;
      this.client = client;
    }
    if (this.client == null) {
      this.client = new Client(this.hosts, this.sslopts);
    }
  }

  Etcd.prototype.set = function(key, value, options, callback) {
    var opt, ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    opt = this._prepareOpts("keys/" + this._stripSlashPrefix(key), "/v2", value, options);
    return this.client.put(opt, callback);
  };

  Etcd.prototype.setSync = function(key, value, options) {
    if (options == null) {
      options = {};
    }
    options.synchronous = true;
    return this.set(key, value, options);
  };

  Etcd.prototype.get = function(key, options, callback) {
    var opt, ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    opt = this._prepareOpts("keys/" + this._stripSlashPrefix(key), "/v2", null, options);
    return this.client.get(opt, callback);
  };

  Etcd.prototype.getSync = function(key, options) {
    if (options == null) {
      options = {};
    }
    options.synchronous = true;
    return this.get(key, options);
  };

  Etcd.prototype.create = function(dir, value, options, callback) {
    var opt, ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    opt = this._prepareOpts("keys/" + this._stripSlashPrefix(dir), "/v2", value, options);
    return this.client.post(opt, callback);
  };

  Etcd.prototype.post = Etcd.prototype.create;

  Etcd.prototype.del = function(key, options, callback) {
    var opt, ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    opt = this._prepareOpts("keys/" + this._stripSlashPrefix(key), "/v2", null, options);
    return this.client["delete"](opt, callback);
  };

  Etcd.prototype["delete"] = Etcd.prototype.del;

  Etcd.prototype.delSync = function(key, options) {
    if (options == null) {
      options = {};
    }
    options.synchronous = true;
    return this.del(key, options);
  };

  Etcd.prototype.mkdir = function(dir, options, callback) {
    var ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    options.dir = true;
    return this.set(dir, null, options, callback);
  };

  Etcd.prototype.mkdirSync = function(dir, options) {
    if (options == null) {
      options = {};
    }
    options.synchronous = true;
    return this.mkdir(dir, options);
  };

  Etcd.prototype.rmdir = function(dir, options, callback) {
    var ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    options.dir = true;
    return this.del(dir, options, callback);
  };

  Etcd.prototype.rmdirSync = function(dir, options) {
    if (options == null) {
      options = {};
    }
    options.synchronous = true;
    return this.rmdir(dir, options);
  };

  Etcd.prototype.compareAndSwap = function(key, value, oldvalue, options, callback) {
    var ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    if (options == null) {
      options = {};
    }
    options.prevValue = oldvalue;
    return this.set(key, value, options, callback);
  };

  Etcd.prototype.testAndSet = Etcd.prototype.compareAndSwap;

  Etcd.prototype.compareAndDelete = function(key, oldvalue, options, callback) {
    var ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    if (options == null) {
      options = {};
    }
    options.prevValue = oldvalue;
    return this.del(key, options, callback);
  };

  Etcd.prototype.testAndDelete = Etcd.prototype.compareAndDelete;

  Etcd.prototype.raw = function(method, key, value, options, callback) {
    var opt, ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    opt = this._prepareOpts(key, "", value, options);
    return this.client.execute(method, opt, callback);
  };

  Etcd.prototype.watch = function(key, options, callback) {
    var ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    if (options == null) {
      options = {};
    }
    options.wait = true;
    return this.get(key, options, callback);
  };

  Etcd.prototype.watchIndex = function(key, index, options, callback) {
    var ref;
    ref = this._argParser(options, callback), options = ref[0], callback = ref[1];
    if (options == null) {
      options = {};
    }
    options.waitIndex = index;
    return this.watch(key, options, callback);
  };

  Etcd.prototype.watcher = function(key, index, options) {
    if (index == null) {
      index = null;
    }
    if (options == null) {
      options = {};
    }
    return new Watcher(this, key, index, options);
  };

  Etcd.prototype.machines = function(callback) {
    var opt;
    opt = this._prepareOpts("keys/_etcd/machines");
    return this.client.get(opt, callback);
  };

  Etcd.prototype.getHosts = function() {
    return _.clone(this.hosts);
  };

  Etcd.prototype.leader = function(callback) {
    var opt;
    opt = this._prepareOpts("leader");
    return this.client.get(opt, callback);
  };

  Etcd.prototype.leaderStats = function(callback) {
    var opt;
    opt = this._prepareOpts("stats/leader");
    return this.client.get(opt, callback);
  };

  Etcd.prototype.selfStats = function(callback) {
    var opt;
    opt = this._prepareOpts("stats/self");
    return this.client.get(opt, callback);
  };

  Etcd.prototype.version = function(callback) {
    var opt;
    opt = this._prepareOpts("version", "");
    return this.client.get(opt, callback);
  };

  Etcd.prototype._stripSlashPrefix = function(key) {
    return key.replace(/^\//, '');
  };

  Etcd.prototype._prepareOpts = function(path, apiVersion, value, allOpts) {
    var clientOptions, opt, queryString, serverprotocol;
    if (apiVersion == null) {
      apiVersion = "/v2";
    }
    if (value == null) {
      value = null;
    }
    if (allOpts == null) {
      allOpts = {};
    }
    serverprotocol = this.sslopts != null ? "https" : "http";
    queryString = _.omit(allOpts, 'maxRetries', 'synchronous');
    clientOptions = _.pick(allOpts, 'maxRetries');
    return opt = {
      path: apiVersion + "/" + path,
      serverprotocol: serverprotocol,
      json: true,
      qs: queryString,
      clientOptions: clientOptions,
      synchronous: allOpts.synchronous,
      form: value != null ? {
        value: value
      } : void 0,
      agentOptions: this.sslopts != null ? this.sslopts : void 0
    };
  };

  Etcd.prototype._argParser = function(options, callback) {
    if (typeof options === 'function') {
      return [{}, options];
    } else {
      return [options, callback];
    }
  };

  return Etcd;

})();

exports = module.exports = Etcd;
